### ISSUE TITLE
Undisciplined #if directives cause false syntax errors due to caching of "if" token

### ISSUE THREAD

**Issue by mayshukla** [ualberta-smr/varclang#44]
Example:

```cpp
struct myStruct
#if 1 > 0
#endif
{
};
```

This is the output with the -ast-print option:
```
clara_failed_assert_if.cpp:13:5: error: token is not a valid binary operator in a
      preprocessor subexpression
#if 1 > 0
 ~~ ^
clara_failed_assert_if.cpp:12:16: error: expected ';' after struct
struct myStruct
               ^
               ;
clara_failed_assert_if.cpp:16:1: error: extraneous closing brace ('}')
};
^
struct myStruct; // In context: True
; // In context: True
3 errors generated.
```

Note that the `if` has been included in the conditional expression and the preprocessor has reported a syntax error since it treats `if` as an identifier and is expecting an operator in place of `1`. Also, the curly braces have been missed (don't appear in the ast).

From what I can tell, the `if` token is being cached when `EnableBacktrackAtThisPos` is called at the beginning of `HandleIfDirective`. Then, after backtracking, the `if` token is returned from the cache. Normally, the `if` token has already been consumed and is not cached when `EnableBacktrackAtThisPos`.

**Comment by mayshukla** [comment]
Currently, this issue pertains to branch `if-defined`

**Comment by mayshukla** [comment]
Update: this bug also occurs in the system `utility` header file in the following snippet:

```cpp
   using make_integer_sequence
#if _GLIBCXX_USE_MAKE_INTEGER_SEQ
      = __make_integer_seq<integer_sequence, _Tp, _Num>;
#else
      = integer_sequence<_Tp, __integer_pack(_Num)...>;
#endif
```

This is the error message:

```
/usr/lib64/gcc/x86_64-pc-linux-gnu/8.3.0/../../../../include/c++/8.3.0/utility:326:5: error: token is not a valid binary operator in a preprocessor subexpression                 
#if _GLIBCXX_USE_MAKE_INTEGER_SEQ                                                                                                                                                 
 ~~ ^
```

I suspect that the caching of the `if` token is happening whenever there are undisciplined annotations. (That is, `#if` directives that occur in the middle of a statement or declaration).

**Comment by mayshukla** [comment]
Here is my explanation of the problem:

Background info:
* The parser uses the method `Parser::NextToken` to "peek" ahead at the next token without consuming it
  * This method calls `Preprocessor::LookAhead` which either returns a token from the cache, or calls `Preprocessor::PeekAhead`
* `Preprocessor::PeekAhead` is used to peek at a token that has not been lexed yet (i.e. is beyond the cache)
  * `Preprocessor::PeekAhead` loops as many times as necessary, depending on how many tokens it is asked to peek ahead
    * In this loop it pushes a blank token to `CachedTokens` by calling `CachedTokens.push_back(Token());`
	* It then calls `Lex(CachedTokens.back())`. This passes the blank token by reference to the `Lex` method. The `Lex` method lexes a token "into" this blank token.
* The member variable `Preprocessor::CachedLexPos` keeps track of the index in the `CachedTokens` vector of the next token that `Lex` should return.
  * This is only if the preprocessor is is caching mode. Otherwise, `Lex` doesn't return tokens from the cache.
  * When `EnableBacktrackAtThisPos` is called, the current value of `CachedLexPos` is pushed onto `BacktrackPositions`
* `PeekAhead` does not increment `CachedLexPos`
  * This is because `PeekAhead` lets caller "peek" at the next token without consuming it.

At a high level, this bug occurs when there are undisciplined annotations because `HandleIfDirective` gets called within the call to `Lex` within `PeekAhead`.

Here is an example of a program that causes the bug:

```cpp
int i
#if 0
= 1;
#else
= 0;
#endif
```

* After consuming the `i` in `int i`, the parser calls `NextToken` in `ParseDeclarationInternal` to see if the next token is `::` (since it needs to handle a special case if the next token is `::`).
* `Parser::NextToken` calls `Preprocessor::LookAhead` which calls `Preprocessors::PeekAhead`
* `PeekAhead` pushes a blank token to `CachedTokens`. Then it calls `Lex(CachedTokens.back())`. Note that the token argument to `Lex` is passed by reference.
  * Note: since `PeekAhead` does not modify `CachedLexPos`, `CachedLexPos` points at the token that was just pushed back. (CachedLexPos normally points just past the end of `CachedTokens`)
* In the call to `Lex`:
  * `Lex` lexes a `#`. Since it was passed by reference, the token that was pushed to `CachedTokens` becomes a `#` token.
  * Since `Lex` just lexed a `#` it calls `HandleDirective` and passes the same token by reference.
* `HandleDirective` needs to know what kind of directive it is looking at. Therefore it must lex the next token after the `#`. To do this it calls `LexUnexpandedToken` and passes the same token. After this call, the token that was pushed to `CachedTokens` becomes an `if` token.
* `HandleDirective` then calls `HandleIfDirective`. At the beginning of `HandleIfDirective`, `EnableBacktrackAtThisPos` is called. This sets the backtrack position to point at the token that was pushed to `CachedTokens` since `CachedLexPos` points there (see background info).
* `HandleDirective` calls `TryParsePresenceCondition` which returns `nullptr` because it fails.
* Since `TryParsePresenceCondition` failed, `HandleDirective` calls `Backtrack` before calling `EvaluateDirectiveExpression` to evaluate the preprocessor expression normally.
* The backtrack position points at the token that was pushed to `CachedTokens`. This token is now an `if` token.
* `EvaluateDirectiveExpression` sees `if 0` and reports an error because it thinks the directive was `#if if 0`

Note that this only happens when `HandleIfDirective` gets called within `PeekAhead`. That's because `PeekAhead` creates a new token without advancing `CachedLexPos`.

**Comment by mayshukla** [comment]
Also note that the token returned by `PeekAhead` (back of CachedTokens) must be the next token after preprocessor directives are dealt with. For example, in this program:

```cpp
int i
#if 0
= 1;
#else
= 0;
#endif
```

Calling `PeekAhead` when at `i` should return `=`. `PeekAhead` returns the last token in `CachedTokens`. This means that just before `PeekAhead` returns, the last token in `CachedTokens` should be `=`.

This normally works fine since `PeekAhead` calls `Lex(CachedTokens.back())`. `Lex` will loop until it returns a proper (i.e. not a preprocessor) token. However, since I am calling `EraseCachedTokens`, all the tokens starting at the `if` to the end of `CachedTokens` get deleted in `HandleIfDirective`. Therefore, when `Lex` has returned and we get to the end of `PeekAhead` the token at the end of `CachedTokens` is whatever was in the cache before we began lexing the `#if` directive. In this case, this happens to be the `i` token.

**Comment by mayshukla** [comment]
Fixed in commit e2b4c7b6ffbbc37c34bac13e6b0151f593611433

My solution was to temporarily create a new `CachedTokens` vector inside `HandleIfDirective`. This prevents any caching that is done by `TryParsePresenceCondition` from interfering with caching done elsewhere.

Also note that we no longer have to call `EraseCachedTokens` after `TryParsePresenceCondition` since any tokens that were cached are put into the temporary vector.
